### PR TITLE
feat(BlockPattern): let autoBuild use itemHandler instead of PlayerInv

### PR DESCRIPTION
**Pull request by DancingSnow0517**
## What
let autoBuild use itemHandler that can use item from Shulker Box or other like item

### CODE REVIEW

File: src/main/java/com/gregtechceu/gtceu/api/pattern/BlockPattern.java [hunk 1]
```diff
@@ -642,4 +647,27 @@ private BlockPos setActualRelativeOffset(int x, int y, int z, Direction facing,
         }
         return new BlockPos(c1[0], c1[1], c1[2]);
     }
+
+    @Nullable
+    private static Pair<Integer, IItemHandler> getMatchStackWithHandler(
+                                                                        List<ItemStack> candidates,
+                                                                        LazyOptional<IItemHandler> cap) throws RuntimeException {
+        IItemHandler handler = cap.orElseThrow(() -> new RuntimeException("No handler available"));
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/pattern/BlockPattern.java]: I don't think this should throw, but just do an early exit by returning null.

File: src/main/java/com/gregtechceu/gtceu/api/pattern/BlockPattern.java [hunk 2]
```diff
@@ -302,16 +308,15 @@ public void autoBuild(Player player, MultiblockState worldState) {
...
                             if (!player.isCreative()) {
-                                for (ItemStack itemStack : player.getInventory().items) {
-                                    if (candidates.stream().anyMatch(
-                                            candidate -> ItemStack.isSameItemSameTags(candidate, itemStack)) &&
-                                            !itemStack.isEmpty() && itemStack.getItem() instanceof BlockItem) {
-                                        found = itemStack.copy();
-                                        originalItemStack = itemStack;
-                                        break;
-                                    }
+                                var founded = getMatchStackWithHandler(candidates,
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/pattern/BlockPattern.java]: name the field `foundHandler` instead of `founded`

**Review by screret — CHANGES_REQUESTED** [review]

File: src/main/java/com/gregtechceu/gtceu/api/pattern/BlockPattern.java (hunk 1 above)
Comment (DancingSnow0517) [reply src/main/java/com/gregtechceu/gtceu/api/pattern/BlockPattern.java]: fixed

File: src/main/java/com/gregtechceu/gtceu/api/pattern/BlockPattern.java (hunk 2 above)
Comment (DancingSnow0517) [reply src/main/java/com/gregtechceu/gtceu/api/pattern/BlockPattern.java]: renamed

**Review by screret — APPROVED** [review]

**Pull request merged**